clients/application-kafka/src/test/java/org/eclipse/hono/application/client/kafka/impl/KafkaApplicationClientImplTest.java

```diff
@@ -0,0 +1,232 @@
+import org.eclipse.hono.client.kafka.consumer.KafkaConsumerConfigProperties;
+import org.eclipse.hono.kafka.test.KafkaClientUnitTestHelper;
+import org.eclipse.hono.test.VertxMockSupport;
+import org.eclipse.hono.util.MessageHelper;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.junit5.Timeout;
+import io.vertx.junit5.VertxExtension;
+import io.vertx.junit5.VertxTestContext;
+import io.vertx.kafka.client.consumer.KafkaConsumer;
+
+/**
+ * Verifies behavior of {@link KafkaApplicationClientImpl}.
+ */
+@ExtendWith(VertxExtension.class)
+@Timeout(value = 5, timeUnit = TimeUnit.SECONDS)
+public class KafkaApplicationClientImplTest {
+    private static final String PARAMETERIZED_TEST_NAME_PATTERN = "{displayName} [{index}]; parameters: {argumentsWithNames}";
+    private KafkaApplicationClientImpl client;
+    private MockConsumer<String, Buffer> mockConsumer;
+    private KafkaConsumer<String, Buffer> kafkaConsumer;
+    private KafkaConsumerConfigProperties consumerConfig;
+    private String tenantId;
+    private String deviceId;
+    private String correlationId;
+
+    static Stream<Type> messageTypes() {
+        return Stream.of(
+                Type.TELEMETRY,
+                Type.EVENT,
+                Type.COMMAND_RESPONSE);
+    }
+
+    /**
+     *
+     * Sets up fixture.
+     *
+     * @param vertx The vert.x instance to use.
+     */
+    @BeforeEach
+    void setUp(final Vertx vertx) {
+        final KafkaProducerConfigProperties producerConfig = new KafkaProducerConfigProperties();
+        final MockProducer<String, Buffer> mockProducer = KafkaClientUnitTestHelper.newMockProducer(true);
+        final CachingKafkaProducerFactory<String, Buffer> producerFactory = KafkaClientUnitTestHelper
+                .newProducerFactory(mockProducer);
+
+        tenantId = UUID.randomUUID().toString();
+        deviceId = UUID.randomUUID().toString();
+        correlationId = UUID.randomUUID().toString();
+
+        mockConsumer = new MockConsumer<>(OffsetResetStrategy.EARLIEST);
+        kafkaConsumer = KafkaConsumer.create(vertx, mockConsumer);
+
+        consumerConfig = new KafkaConsumerConfigProperties();
+        consumerConfig.setConsumerConfig(Map.of("client.id", "application-test-consumer"));
+        producerConfig.setCommonClientConfig(Map.of("client.id", "application-test-sender"));
+
+        client = new KafkaApplicationClientImpl(vertx, consumerConfig, producerFactory, producerConfig) {
+
+            //Override to use the mocked Kafka consumer
+            @Override
+            Future<MessageConsumer> createKafkaBasedDownstreamMessageConsumer(final String tenantId, final Type type,
```

Comment:
If I understand the code correctly, the underlying problem is the hard wired dependency on *KafkaConsumer.create(Vertx, Consumer)* which makes it very hard to test. How about adding something like
*KafkaApplicationClientImpl.setKafkaConsumerFactory(Function<Consumer, KafkaConsumer>)* which can be used by the test to set a function that produces a mocked KafkaConsumer?

Reply:
Having some factory method is a good idea. I think instead of passing a `Consumer` we can directly supply a `KafkaConsumer` as below.

```   
 KafkaApplicationClientImpl.setKafkaConsumerFactory(final Supplier<KafkaConsumer<String, Buffer>> kafkaConsumerSupplier)
```

Reply:
I have updated and pushed the changes, please take a look.